### PR TITLE
Add `ptr::{str_from_raw_parts, str_from_raw_parts_mut}` functions

### PR DESCRIPTION
**Pull request by WaffleLapkin**
TL;DR: this PR adds the following API:
```rust
// core::ptr
pub const fn str_from_raw_parts(data: *const u8, len: usize) -> *const str;
pub const fn str_from_raw_parts_mut(data: *mut u8, len: usize) -> *mut str;

impl NonNull<str> {
    pub const fn str_from_raw_parts(data: NonNull<u8>, len: usize) -> Self;
}
```

The functions are under feature gate `str_from_raw_parts` and are similar to
`slice_from_raw_parts`, `slice_from_raw_parts_mut` and `NonNull::slice_from_raw_parts`.

---

These methods were originally proposed as a part of #85816 and there were some [concerns](https://github.com/rust-lang/rust/pull/85816#issuecomment-851395458) about the specific design we should use for `str` pointer construction.

Some alternative designs:

- "Blessed cast"
   ```rust
   // core::ptr
   pub const fn str_from_raw_utf8(raw: *const [u8]) -> *const str;
   pub const fn str_from_raw_utf8_mut(raw: *mut [u8]) -> *mut str;
   // Also NonNull?
   ```
   Pros: similar to already existing `&str` construction (`str::from_utf8`)
   Cons: forces to use 2 functions (`str_from_raw_utf8(slice_from_raw_parts(ptr, len))`) to convert `ptr, len` into `*const str`
-  Add `str::from_raw_parts` too (addition to the current design of this PR)
   ```rust
   // core::str
   pub const unsafe fn from_raw_parts<'a>(ptr: *const u8, len: usize) -> &'a str;
   pub const unsafe fn from_raw_parts_mut<'a>(ptr: *mut u8, len: usize) -> &'a mut str;
   ```
   Pros: similar to slices
   Cons: adds a function to `str` which can be replaced with `slice::from_raw_parts` and `str::from_utf8_unchecked` (in this case having two functions is probably better, as it splits safety conditions into smaller parts)
- Do nothing, use `ptr::from_raw_parts`
  Pros: doesn't add anything new, reuses existing API
  Cons: less clear/longer stabilization path

### CODE REVIEW

**Review by dtolnay — COMMENTED** [review]
I would prefer not to add these. In the long term, `ptr::from_raw_parts` should be fine for this use case. In the short term, it sounded like the only immediate motivation for adding these APIs was:

> https://github.com/rust-lang/rust/pull/85816#issuecomment-851658408
> it seemed weird to have methods on `*const str` without a convenient way to create such pointers

This isn't compelling to me because people are successfully and reasonably conveniently working with `*const str` in real code today; they aren't the ones asking for this. A few examples:

#### `as` cast from `*const [u8]`:

https://github.com/rust-lang/rust/blob/23f69235ad2eb9b44ac1a55eeaa3f9b484d9de4a/library/alloc/src/rc.rs#L1818-L1819

#### `as` cast from `&str`:

https://github.com/rust-lang/rust/blob/23f69235ad2eb9b44ac1a55eeaa3f9b484d9de4a/compiler/rustc_span/src/symbol.rs#L1767-L1772

**Comment by WaffleLapkin** [comment]
@dtolnay that makes sense. I didn't know the `*const [u8]` can be cast to `*const str`, I thought that `as` only worked for pointers to sized types...

Considering how easy it is to construct `*const str` and `ptr::from_raw_parts`, this proposal indeed doesn't seem to be useful enough. Closing as such.
